clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+            *      pid of the abort index entry, and the incoming offset is no smaller than the abort index offset,
+            *      this means that the entry has been aborted. Add the pid to the aborted pids set, and remove
+            *      the entry from the abort index.
+            */
+            if (isolationLevel == IsolationLevel.READ_COMMITTED) {
+                FetchResponse.AbortedTransaction nextAbortedTransaction = abortedTransactions.peek();
+                if (abortedPids.contains(batch.producerId())
+                        || (nextAbortedTransaction != null && nextAbortedTransaction.producerId == batch.producerId() && nextAbortedTransaction.firstOffset <= batch.baseOffset())) {
+                    if (abortedPids.contains(batch.producerId()) && containsAbortMarker(batch)) {
+                        abortedPids.remove(batch.producerId());
+                    } else if (nextAbortedTransaction != null && nextAbortedTransaction.producerId == batch.producerId() && nextAbortedTransaction.firstOffset <= batch.baseOffset()) {
+                        abortedPids.add(batch.producerId());
+                        abortedTransactions.remove(nextAbortedTransaction);
+                    }
+                    log.trace("Skipping aborted record with pid {} and base offset {}", batch.producerId(), batch.baseOffset());
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private Queue<FetchResponse.AbortedTransaction> abortedTransactions(FetchResponse.PartitionData partition) {
+            PriorityQueue<FetchResponse.AbortedTransaction> abortedTransactions = null;
+            if (partition.abortedTransactions != null && 0 < partition.abortedTransactions.size()) {
+                abortedTransactions = new PriorityQueue<>(
+                        partition.abortedTransactions.size(),
+                        new Comparator<FetchResponse.AbortedTransaction>() {
+                            @Override
+                            public int compare(FetchResponse.AbortedTransaction o1, FetchResponse.AbortedTransaction o2) {
+                                return (int) o1.firstOffset - (int) o2.firstOffset;
+                            }
+                        }
+                );
+                abortedTransactions.addAll(partition.abortedTransactions);
+            } else {
+                abortedTransactions = new PriorityQueue<>();
+            }
+            return abortedTransactions;
+        }
+
+        private boolean containsAbortMarker(RecordBatch batch) {
+            Iterator<Record> batchIterator = batch.iterator();
+            Record firstRecord = batchIterator.hasNext() ? batchIterator.next() : null;
+            if (firstRecord != null && batchIterator.hasNext())
+                throw new CorruptRecordException("A RecordBatch containing a control message contained more than one message.");
```

Comment:
nit: "RecordBatch" -> "record batch"? Also, may want to include a bit more detail here such as the partition and the offset.